File: frontend/mobile_app/lib/features/connection/connection_controller.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../../core/realtime/realtime_service.dart';
import '../../models/session.dart';
import '../../models/ws_event.dart';

enum ConnectionStatus { disconnected, connecting, connected, error }

/// Manages login + QR pairing to a display and the resulting [SessionInfo].
///
/// Flow (mirrors REQUIREMENTS §5/§7): scan QR → parse [PairingInfo] → emit
/// `pair {token}` → (server binds) → `paired {sessionId, displayId}`. In the POC
/// the "server" is simulated locally; swapping [RealtimeService] for the real
/// LAN client makes this real with no UI change.
class ConnectionController extends ChangeNotifier {
  ConnectionController(this._realtime) {
    _sub = _realtime.events.listen(_onEvent);
  }

  final RealtimeService _realtime;
  StreamSubscription<WsEvent>? _sub;

  ConnectionStatus status = ConnectionStatus.disconnected;
  SessionInfo? session;
  String? error;

  /// True when a real WebSocket link to the display was established (vs. an
  /// offline demo fallback).
  bool liveLink = false;

  /// Idle-timeout warning surfaced from the display; prompts the associate to
  /// keep the session alive.
  bool idleWarning = false;
  int idleSecondsLeft = 0;

  bool get isConnected => status == ConnectionStatus.connected;

  void _onEvent(WsEvent e) {
    switch (e.type) {
      case WsEventType.sessionWarning:
        idleWarning = true;
        idleSecondsLeft = e.secondsLeft ?? 0;
        notifyListeners();
      case WsEventType.sessionEnd:
      case WsEventType.sessionTimeout:
        _reset();
      default:
        break;
    }
  }

  /// Associate confirms they are still present; cancels the idle warning.
  void keepAlive() {
    idleWarning = false;
    _realtime.emit(
      WsEvent(
        type: WsEventType.keepAlive,
        sessionId: session?.sessionId,
        senderRole: SenderRole.salesperson,
      ),
    );
    notifyListeners();
  }

  void _reset() {
    idleWarning = false;
    liveLink = false;
    session = null;
    status = ConnectionStatus.disconnected;
    notifyListeners();
  }

  /// End the session (e.g. mobile idle timeout): clear the display, close the
  /// live transport so the server frees the screen (returns it to idle + QR),
  /// and drop back to the connect screen.
  Future<void> endSession() async {
    _realtime.emit(
      WsEvent(
        type: WsEventType.hideProduct,
        sessionId: session?.sessionId,
        senderRole: SenderRole.salesperson,
      ),
    );
    await _realtime.closeTransport();
    _reset();
  }

  Future<void> connectFromQr(String raw, {Salesperson? salesperson}) async {
    final PairingInfo? info = PairingInfo.tryParse(raw);
    if (info == null) {
      error = 'That QR code is not a valid display pairing code.';
      status = ConnectionStatus.error;
      notifyListeners();
      return;
    }
    await connect(info, salesperson: salesperson);
  }

  Future<void> connect(PairingInfo info, {Salesperson? salesperson}) async {
    error = null;
    status = ConnectionStatus.connecting;
    notifyListeners();

    final Salesperson person =
        salesperson ?? const Salesperson(id: 's0', name: 'Associate');

    // Attempt a real WebSocket link to the display's LAN server. If it fails
    // (e.g. the demo display), we fall back to an offline session so the app
    // stays fully usable.
    liveLink = await _realtime.connect(info.wsUri);

    _realtime.emit(
      WsEvent(
        type: WsEventType.pair,
        senderRole: SenderRole.salesperson,
        payload: <String, dynamic>{
          'token': info.token,
          'salespersonId': person.id,
        },
      ),
    );

    // Brief binding delay (real server acknowledges; fallback simulates).
    await Future<void>.delayed(const Duration(milliseconds: 700));

    session = SessionInfo(
      sessionId: 'sess_${info.token}',
      displayId: 'disp_${info.host}',
      salesperson: person,
    );
    status = ConnectionStatus.connected;
    idleWarning = false;

    _realtime.emit(
      WsEvent(
        type: WsEventType.connectScreen,
        sessionId: session!.sessionId,
        senderRole: SenderRole.salesperson,
        payload: <String, dynamic>{'salespersonName': person.name},
      ),
    );
    notifyListeners();
  }

  void disconnect() {
    if (session != null) {
      _realtime.emit(
        WsEvent(
          type: WsEventType.disconnectScreen,
          sessionId: session!.sessionId,
          senderRole: SenderRole.salesperson,
        ),
      );
    }
    session = null;
    status = ConnectionStatus.disconnected;
    notifyListeners();
  }

  @override
  void dispose() {
    _sub?.cancel();
    super.dispose();
  }
}

File: frontend-rn/mobile_app/src/features/connection/connectionController.ts
import { Listenable } from '../../core/listenable';
import { PairingInfo, Salesperson, SessionInfo } from '../../models/session';
import { WsEvent, WsEventType } from '../../models/wsEvent';
import { RealtimeService } from '../../realtime/realtimeService';

export enum ConnectionStatus {
  disconnected = 'disconnected',
  connecting = 'connecting',
  connected = 'connected',
  error = 'error',
}

/// Manages QR pairing to a display and the resulting SessionInfo. Ported 1:1
/// from `ConnectionController`.
export class ConnectionController extends Listenable {
  status = ConnectionStatus.disconnected;
  session: SessionInfo | null = null;
  error: string | null = null;
  liveLink = false;
  idleWarning = false;
  idleSecondsLeft = 0;

  private sub: { cancel: () => void } | null = null;

  constructor(private readonly realtime: RealtimeService) {
    super();
    this.sub = this.realtime.events.listen((e) => this.onEvent(e));
  }

  get isConnected(): boolean {
    return this.status === ConnectionStatus.connected;
  }

  private onEvent(e: WsEvent): void {
    switch (e.type) {
      case WsEventType.sessionWarning:
        this.idleWarning = true;
        this.idleSecondsLeft = e.secondsLeft ?? 0;
        this.notifyListeners();
        break;
      case WsEventType.sessionEnd:
      case WsEventType.sessionTimeout:
        this.reset();
        break;
      default:
        break;
    }
  }

  keepAlive(): void {
    this.idleWarning = false;
    this.realtime.emit(
      new WsEvent({ type: WsEventType.keepAlive, sessionId: this.session?.sessionId, senderRole: 'salesperson' }),
    );
    this.notifyListeners();
  }

  private reset(): void {
    this.idleWarning = false;
    this.liveLink = false;
    this.session = null;
    this.status = ConnectionStatus.disconnected;
    this.notifyListeners();
  }

  async endSession(): Promise<void> {
    this.realtime.emit(
      new WsEvent({ type: WsEventType.hideProduct, sessionId: this.session?.sessionId, senderRole: 'salesperson' }),
    );
    await this.realtime.closeTransport?.();
    this.reset();
  }

  async connectFromQr(raw: string, opts?: { salesperson?: Salesperson }): Promise<void> {
    const info = PairingInfo.tryParse(raw);
    if (info == null) {
      this.error = 'That QR code is not a valid display pairing code.';
      this.status = ConnectionStatus.error;
      this.notifyListeners();
      return;
    }
    await this.connect(info, opts);
  }

  async connect(info: PairingInfo, opts?: { salesperson?: Salesperson }): Promise<void> {
    this.error = null;
    this.status = ConnectionStatus.connecting;
    this.notifyListeners();

    const person = opts?.salesperson ?? new Salesperson({ id: 's0', name: 'Associate' });

    // Attempt a real WebSocket link; fall back to an offline session on failure.
    this.liveLink = (await this.realtime.connect?.(info.wsUrl)) ?? false;

    this.realtime.emit(
      new WsEvent({
        type: WsEventType.pair,
        senderRole: 'salesperson',
        payload: { token: info.token, salespersonId: person.id, salespersonName: person.name },
      }),
    );

    // Brief binding delay (real server acknowledges; fallback simulates).
    await new Promise<void>((r) => setTimeout(r, 700));

    this.session = new SessionInfo({
      sessionId: `sess_${info.token}`,
      displayId: `disp_${info.host}`,
      salesperson: person,
    });
    this.status = ConnectionStatus.connected;
    this.idleWarning = false;

    this.realtime.emit(
      new WsEvent({
        type: WsEventType.connectScreen,
        sessionId: this.session.sessionId,
        senderRole: 'salesperson',
        payload: { salespersonName: person.name },
      }),
    );
    this.notifyListeners();
  }

  disconnect(): void {
    if (this.session != null) {
      this.realtime.emit(
        new WsEvent({ type: WsEventType.disconnectScreen, sessionId: this.session.sessionId, senderRole: 'salesperson' }),
      );
    }
    this.session = null;
    this.status = ConnectionStatus.disconnected;
    this.notifyListeners();
  }

  dispose(): void {
    this.sub?.cancel();
  }
}
